doc/design/dynamic_tls.md

```diff
@@ -0,0 +1,66 @@
+# Dynamic TLS Certs for etcd Cluster in etcd Operator
+
+Author: Eric Chiang(eric.chiang@coreos.com), Hongchao Deng(hongchao.deng@coreos.com)
+
+Motivation:
+- Simplify TLS cert generation for etcd cluster.
+- Self-hosted etcd wants to depend on Node IP, not hostnames.
+
+Requirements:
+- Dynamically generate certs per etcd pod
+- Able to use host IP in self hosted.
+
+### Option 1: K8S CSR, generate on etcd pod init container
+
+[Kubernetes certificates API][k8s_csr] provides the ability to send and approve CSR.
+
+Workflow:
+- etcd operator creates an etcd pod.
+  There is a separate service account for etcd pod to access  CSR API.
+- etcd pod first runs an init container to send a CSR and then wait for approval.
+- An external process should verify the CSR and approve it if correct.
+- Once approved, etcd pod gets the certs and bootstraps.
+
+Pros:
+- Hooks into existing k8s.
+- Delegates CA signing cert management. Not touching the signing certs and keys directly.
+
+Cons:
+- Share the same CA signing cert + key as k8s.
+- Requires etcd to be trusted by k8s cluster CA (probably only appropriate for self-hosted etcd).
+- Requires additional admin setup: approver, etcd pod service account.
+
+
+### Option 2: CA signing cert + key, generate on init etcd pod init container
+
+Add a secret reference to CA signing cert + key in EtcdCluster spec. Use it to generate certs for etcd pod.
+
+Workflow:
+- etcd operator creates an etcd pod with a volume mounted with CA secret.
+- etcd pod first runs an init container to use the CA cert + key to generate certs in an emptyDir.
+- etcd container then uses the generated certs in bootstrap.
+
+Pros:
+- Each etcd cluster has its own CA.
+
+Cons:
+- Inappropriate to give etcd pod access to the CA key?
```

Comment:
Answer: yes, it's inappropriate to give the etcd pod the CA key.

I don't think we should include this as an option.

Reply:
:+1: That was my point during yesterday's meeting where we discussed this.

> I don't think we should include this as an option.

SGTM